src/nf/nf_network_submodule.f90

```diff
@@ -369,6 +369,49 @@ module subroutine print_info(self)
   end subroutine print_info
 
 
+  pure module function get_num_params(self) result(num_params)
+    class(network), intent(in) :: self
+    integer :: n, num_params
+
+    num_params = 0
+
+    do n = 1, size(self % layers)
+      num_params = num_params + self % layers(n) % get_num_params()
+    end do
+
+  end function get_num_params
+
+
+  pure module subroutine get_params(self, params)
+    class(network), intent(in) :: self
+    real, allocatable, intent(in out) :: params(:)
+    integer :: n
+
+    do n = 1, size(self % layers)
+      call self % layers(n) % get_params(params)
+    end do
+
+  end subroutine get_params
+
+
+  module subroutine set_params(self, params)
+    class(network), intent(in out) :: self
+    real, intent(in) :: params(:)
+    integer :: n, consumed
+
+    ! Check that the number of parameters is correct.
+    if (size(params) .ne. self % get_num_params()) then
+      error stop 'network % set_params: number of parameters does not match.'
+    end if
+
+    consumed = 0
+    do n = 1, size(self % layers)
+      consumed = consumed + self % layers(n) % set_params(params(consumed + 1:))
+    end do
```

Comment:
@jvo203 Is `consumed` here simply because `layer % set_params` is currently a function?

Reply:
Yes, set_params is a function right now. But it could be made a subroutine, with the "consumed" as an "inout" parameter updated by each layer, i.e. ```
```
self%layers(n)%set_params(params, consumed)
```

There is more than one way of doing it. I just picked one but surely you can change it in the final version!

Reply:
Rather than passing all remaining params to each layer and having each layer report back how many it consumed, what do you think about this alternative approach:

In `network % set_params`, which receives the full set of parameters from the user, each iteration determines the start and end index of `params` to pass to `layer % set_params`. Something like this (not tested but should give an idea):

```fortran
    ...
    integer :: nstart, nend
    ...

    nstart = 1
    do n = 1, size(self % layers)
      nend = nstart + self % layers(n) % get_num_params() - 1
      call self % layers(n) % set_params(params(nstart:nend))
      nstart = nend + 1
    end do
```
and in `layer % set_params` (here a subroutine), we'd check that the size of input `params` match the number of params in the layer and `error stop` if not.

I find this a bit cleaner and easier to follow since the `params` slicing is kept to a higher level (in `network` rather than lower levels).

Reply:
OK, I've tested your suggestion in my forked branch. It works fine as it is, the ```nstart``` and ```nend``` are set properly and the layers with ```#params=0``` are handled correctly too. The ```consumed``` variable has been deleted across all the source files, and ```set_parameters(self, params)``` has been converted from a function to a subroutine.
```
      nstart = 1
      do n = 1, size(self%layers)
         nend = nstart + self%layers(n)%get_num_params() - 1
         call self%layers(n)%set_parameters(params(nstart:nend))
         nstart = nend + 1
      end do
```
My only concern is how FORTRAN handles passing arrays with the ending index smaller than the beginning, i.e. ```params(1:0)```, which would happen for the case of zero-parameter layers like inputs etc. Would it not cause undefined behaviour? Anyway, such "degenerate" ```params(1:0)``` is not actually used, it is only passed and no further action is taken within zero-parameter layers, it is simply being ignored. And no segmentation faults happen, the code works as expected.

There is one more change I made inside ```pure module subroutine get_parameters(self, params)``` from ```nf_network_submodule.f90```. There is now an extra line ```if (allocated(params)) deallocate(params)``` before the ```do loop``` starts. This makes the code fool-proof across various unexpected use scenarios:
```
if (allocated(params)) deallocate(params)

      do n = 1, size(self%layers)
         call self%layers(n)%get_parameters(params)
      end do
```

I will **NOT** be submitting a new Pull Request for all these changes as the my forked repository is already out of sync with yours, and there is an added problem with TAB formatting. So if you like your suggested approach with ```params(nstart:nend)``` in ```set_params()``` then please go ahead and apply changes on your side. My side has already been converted to your new approach!

Reply:
Thank you, this is very helpful! Will incorporate into this PR and add you as co-author.

Reply:
OK. That's great!